### PR TITLE
Propose the creation of best practices and guidelines 

### PR DESCRIPTION
**Pull request by samj1912**
Signed-off-by: Sambhav Kothari <skothari44@bloomberg.net>

[Readable](https://github.com/samj1912/rfcs/blob/guidelines/text/0000-best-practices-and-guidelines.md)

@buildpacks/learning-maintainers  @buildpacks/learning-contributors - please feel free to collaborate on the RFC and edit it :)

### CODE REVIEW

File: text/0000-best-practices-and-guidelines.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+Examples - 
+
+- `Bill-of-materials` needs to be regenerated on each run
+
+
+# What it is
+[what-it-is]: #what-it-is
+
+The `guidelines` would be contained in a new repository called `guidelines` which will be created in the `buildpacks` Github organization. This repository will be maintained by the `Learning team`. The repository should have seperate sections that target different personas and list guidelines and best practices specific to them -
```
Comment (hone) [inline text/0000-best-practices-and-guidelines.md:67]: Would you prefer a separate repo just putting it on `buildpacks/docs` as a section?

File: text/0000-best-practices-and-guidelines.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+- Platform operators
+- Platform implementors
+
+These could be detailed sections or checklists that they can follow to ensure that they are following the best practices set by the guidelines.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this?
+N/A?
```
Comment (hone) [inline text/0000-best-practices-and-guidelines.md:80]: The big drawback I see is more burden/maintenance for the learning team.

**Review by hone — APPROVED** [review]

File: text/0000-best-practices-and-guidelines.md (hunk 1 above)
Comment (samj1912) [reply text/0000-best-practices-and-guidelines.md:68]: Yup - added that as an alternative. From the last learning team sync - others seemed in favor of that as well.

File: text/0000-best-practices-and-guidelines.md (hunk 2 above)
Comment (samj1912) [reply text/0000-best-practices-and-guidelines.md]: Added that as a drawback.

File: text/0000-best-practices-and-guidelines.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+- [Best coding practices](https://en.wikipedia.org/wiki/Best_coding_practices)
+- [Rust API guidelines](https://rust-lang.github.io/api-guidelines/about.html)
+- [PEP8 - Python Style-guide](https://www.python.org/dev/peps/pep-0008/)
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- The exact set of guidelines that should be in the first cut of the repository
+- Topics and details to cover
+- Process for amending and creation of new guidelines
```
Comment (jromero) [inline text/0000-best-practices-and-guidelines.md:108]: This is a tough open question. I'd like to come to some agreement in this forum on how we would accept/reject guidelines. 

- Should it be at the learning maintainers discretion? 
- Should guidelines reviewed by core team members?
- Should guidelines be proposed via RFCs?

File: text/0000-best-practices-and-guidelines.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+# Alternatives
+[alternatives]: #alternatives
+
+## Create a new repository
+
+The guidelines and best-practices could be a separate repository under the buildpacks organization. This would allow the guidelines and best-practices to have a different review and approval process for the `guidelines`.
+
+## Extend the spec
+
+Have an extension to the spec that contains a list of `SHOULD`s and `MAY`s that capture the guidelines. This may however confuse users and may tie up something like the `spec` which is not geared towards end-users to `guidelines` which are supposed to be.
```
Comment (dfreilich) [inline text/0000-best-practices-and-guidelines.md:93]: Definitely against this idea - the spec is hard enough to read without asking people to read it even more for best practices, which should by their nature be easy to understand.

**Review by dfreilich — APPROVED** [review]
In general, I think this impetus is great. We had discussed having FAQs in the docs for various tools, and I think this can fit in that same general category of non-intuitive information we want to make sure is understood and can be found. Definitely a fan of this!

**Review by jkutner — APPROVED** [review]

File: text/0000-best-practices-and-guidelines.md (hunk 3 above)
Comment (jromero) [reply text/0000-best-practices-and-guidelines.md:108]: Based on our conversation within the learning team sub-sync, the @buildpacks/learning-maintainers would like to propose that they take ownership via simple pull requests. If we find that it isn't working for one reason or another, we can propose alternative strategies. The primary driver is that we know that we'll likely need to iterate over the strategy and would like to start with the one with lowest effort for both maintainers and contributors.

File: text/0000-best-practices-and-guidelines.md [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+- [Best coding practices](https://en.wikipedia.org/wiki/Best_coding_practices)
+- [Rust API guidelines](https://rust-lang.github.io/api-guidelines/about.html)
+- [PEP8 - Python Style-guide](https://www.python.org/dev/peps/pep-0008/)
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- The exact set of guidelines that should be in the first cut of the repository
+- Topics and details to cover
```
Comment (jromero) [inline text/0000-best-practices-and-guidelines.md:107]: I wonder if creating a [discussion](https://github.com/buildpacks/community/discussions) would be a good way of seeding the initial set of guidelines.

**Review by jromero — APPROVED** [review]

**Comment by jromero** [comment]
FPC based on majority sub-team binding votes. To be merged 4/20.

**Review by yaelharel — APPROVED** [review]
I love this idea and I think @jromero's [suggestion](https://github.com/buildpacks/rfcs/pull/150#discussion_r612510761) to start a discussion is a great way to move this forward.

**Review by ekcasey — APPROVED** [review]

**Pull request merged**